build_retrieval_dataset.py

```diff
@@ -0,0 +1,389 @@
+    """
+    Takes a retrieval image as input and returns a copy of the image, with an homographic transformation applied.
+    By default, the output image has the same shape as the original image.
+    Currently there is only one hardcoded transform available, i.e., not random.
+    :param retrieval_img: the image to transform
+    :param target_shape: the shape of the new image
+    :return: the transformed image
+    """
+    h, w, _ = retrieval_img.shape
+    if target_shape is None:
+        target_shape = retrieval_img.shape[:2]
+    target_h, target_w = target_shape
+
+    pts_src = np.array([[0, 0], [0, w], [h, 0], [h, w]])
+    pts_dst = np.array([[0, 0], [int(0.2*target_w), target_h], [target_w, 0], [int(0.8*target_w), target_h]])
+    h, status = cv.findHomography(pts_src, pts_dst)
+    return cv.warpPerspective(retrieval_img, h, (target_w, target_h))
+
+
+def get_real_coin_size(coin_value, two_euro_reference_size=64):
+    """
+    Returns the size physical size of a coin, assuming a 2€ coin is two_euro_reference_size pixels wide
+    :param coin_value: the value of the coin
+    :param two_euro_reference_size: width/height of a 2€ coin
+    :return: the size of a coin of the given value, with respect to the size of a 2€ coin
+    """
+    # using https://de.wikipedia.org/wiki/Eurom%C3%BCnzen as reference for coin sizes
+    coin_sizes_mm = {200: 2575, 100: 2325, 50: 2425, 20: 2225, 10: 1975, 5: 2125, 2: 1875, 1: 1625}
+    scale_factor = coin_sizes_mm[coin_value] / coin_sizes_mm[200]
+    return int(scale_factor * two_euro_reference_size)
+
+
+def generate_retrieval_image(data_path='data', h=256, w=256, coin_amt_mean=9, do_homographic_transform=True):
+    """
+    Returns an retrieval image with the given height, width and amount of coins; and a set of labels.
+    :param data_path: the location of the dataset
+    :param h: height of the generated image
+    :param w: width of the generated image
+    :param coin_amt_mean: the number of coins
+    :param do_homographic_transform: whether to apply an homographic transform
+    :return: retrieval image, labels
+    """
+    # background = background()    # for now: a random, 1-colored background
+    #background_colors = [(0, 200, 0), (200, 0, 0), (0, 0, 200) ]
+    #background_color = random.choice(background_colors)
+
+    # pic = gen_pic(y,x,background)
+    #retrieval_img = create_bgr_image(h, w, bg=background_color)
+
+    retrieval_img = background_generator.gen_background()
+
+    # label-list: a dictionary mapping coin cent values (integers) to their frequency,
+    # and a 'sum' field containing the accumulated value. We could calculate the sum on the fly from the dictionary too,
+    # but I'm lazy and one additional dictionary entry does not hurt
+    labels = {200: 0, 100: 0, 50: 0, 20: 0, 10: 0, 5: 0, 2: 0, 1: 0, 'sum': 0, 'num_coins': 0} # 'background': background_color}
+
+    # coin_amt = gaussian(coin_amt_mean, var=1)
```

Comment:
Is this working? I think it should work even doing something very simple like:
`
coin_amt = np.ceil(np.random.normal(coin_amt_mean, scale=2))
coin_amt = coin_amt if coin_amt >=1 else 1
`

Reply:
Nope, this is just some pseudo-code snipped that evaded my refactoring :see_no_evil:  well spotted :)
At the moment, the actual coin amount is always set to the mean coin amount (the value that is given as method parameter). I did this mostly for debugging purposes in the beginning. I'm actually slightly in favor of keeping it fixed for now and to add more complexity (varying coin amounts might be more difficult for the coin segmentation step? not sure  about that), but I don't have a strong opinion here.
If you prefer to randomize the number of coins, just tell me and I'll do it :)

Reply:
I think it's better to randomize a little bit the number of coins, maybe by just +-2 or so, just to have a little more testing of the algorithm and to be more realistic. Also, some variety is needed to ensure the retrieval network does not overfit.

Reply:
added like recommended (testing shows, standard devision of 2 seems to give a nice distribution in generated output retrieval dataset)

Reply:
Good job!